LdapForNet/LdapSearchExtensions.cs

```diff
@@ -77,5 +77,62 @@ public static async Task<IList<LdapEntry>> SearchBySidAsync(this ILdapConnection
         {
             return await connection.SearchBySidAsync(LdapUtils.GetDnFromHostname(), sid);
         }
+
+        public static IList<LdapEntry> SearchWithPaging(
+            this ILdapConnection connection,
+            SearchRequest request,
+            int pageSize = 1000)
+        {
+            var results = new List<DirectoryEntry>();
+            var resultRequestControl = new PageResultRequestControl(pageSize);
+            request.Controls.Add(resultRequestControl);
+
+            var response = (SearchResponse)connection.SendRequest(request);
+            results.AddRange(response.Entries);
+
+            PageResultResponseControl pageResultResponseControl;
+            while (true)
+            {
+                pageResultResponseControl = (PageResultResponseControl)response.Controls.FirstOrDefault(_ => _ is PageResultResponseControl);
+                if (pageResultResponseControl == null || pageResultResponseControl.Cookie.Length == 0)
+                {
+                    break;
+                }
+
+                resultRequestControl.Cookie = pageResultResponseControl.Cookie;
+                response = (SearchResponse)connection.SendRequest(request);
+                results.AddRange(response.Entries);
+            }
+            return results.ConvertAll(x => x.ToLdapEntry());
+        }
+
+        public static async Task<IList<LdapEntry>> SearchWithPagingAsync(
+            this ILdapConnection connection,
+            SearchRequest request,
+            int pageSize = 1000,
+            CancellationToken token = default)
+        {
+            var results = new List<DirectoryEntry>();
+            var resultRequestControl = new PageResultRequestControl(pageSize);
+            request.Controls.Add(resultRequestControl);
+
+            var response = (SearchResponse)await connection.SendRequestAsync(request, token);
+            results.AddRange(response.Entries);
+
+            PageResultResponseControl pageResultResponseControl;
+            while (true)
```

Comment:
check cancellationToken instead of true
```
while(!token.IsCancellationRequested)

Reply:
This would result in returning an incomplete result if the token was cancelled before all requests were processed, instead of throwing an exception, which is not a desired behaviour.
With the cancellation token fix PR that was already accepted the SendRequestAsync call will throw an exception if the token was cancelled which is what we would want to happen.